File: bezier-clock/ViewController.m
//
//  ViewController.m
//  bezier-clock
//
//  Created by Alexey Yukin on 11.09.14.
//  Copyright (c) 2014 Simbirsoft Ltd. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (nonatomic, copy) void (^clockUpdateBlock)(BOOL);

@end

@implementation ViewController

- (void) viewDidLoad
{
    [super viewDidLoad];

    // setup appearance
    self.view.backgroundColor = [UIColor blackColor];

    [_digitViews setValue:[UIColor colorWithRed:240.0/255.0 green:100.0/255.0 blue:0.0 alpha:1.0] forKey:@"color"];
    [_digitViews setValue:@(3.0) forKey:@"thickness"];

    // sort views
    _digitViews = [_digitViews sortedArrayUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"tag" ascending:YES]]];

    // update loop
    __weak __typeof(self) weakSelf = self;

    self.clockUpdateBlock = ^(BOOL animated)
    {
        NSDateComponents* components = [[NSCalendar currentCalendar] components:(NSCalendarUnitHour | NSCalendarUnitMinute | NSCalendarUnitSecond) fromDate:[NSDate dateWithTimeIntervalSinceNow:1.0]];

        [weakSelf.digitViews[0] setDigit:(components.hour   / 10) animated:animated];
        [weakSelf.digitViews[1] setDigit:(components.hour   % 10) animated:animated];
        [weakSelf.digitViews[2] setDigit:(components.minute / 10) animated:animated];
        [weakSelf.digitViews[3] setDigit:(components.minute % 10) animated:animated];
        [weakSelf.digitViews[4] setDigit:(components.second / 10) animated:animated];
        [weakSelf.digitViews[5] setDigit:(components.second % 10) animated:animated];

        CFAbsoluteTime time  = CFAbsoluteTimeGetCurrent();
        CFAbsoluteTime delta = ceil(time) - time - [DigitView animationDuration];

        while (delta < 0.0)
        {
            delta = delta + 1.0;
        }

        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, delta * NSEC_PER_SEC), dispatch_get_main_queue(), ^(void)
        {
            if (weakSelf.clockUpdateBlock)
            {
                weakSelf.clockUpdateBlock(YES);
            }
        });
    };

    _clockUpdateBlock(NO);
}

- (BOOL) prefersStatusBarHidden
{
    return YES;
}

@end

File: bezier-clock/ViewController.swift
//
//  SwViewController.swift
//  bezier-clock
//
//  Created by Евгений Губин on 17.09.14.
//  Copyright (c) 2014 Simbirsoft Ltd. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    // Outlet collection - это просто массив (глядите какой модный синтаксис)
    @IBOutlet var digitViews: [DigitView]!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // setup appearance
        self.view.backgroundColor = UIColor.black
        
        // и в отличии от NSArray рассылку сообщений (setValue:) он не поддерживает,
        // поэтому используем for-in и долой строковые константы
        for dw in digitViews {
            dw.color = UIColor(red:240.0/255.0, green:100.0/255.0, blue:0.0, alpha:1.0)
            dw.thickness = 3.0
        }
        
        // sort views
        // вместо NSPredicate используем trailing closures
        digitViews.sort { $0.tag < $1.tag }

        // update loop
        clockUpdate(false)
    }

    // тут подход тот же, что и в Obj-C, но инициализация ленивая синтаксически,
    // а тип объявлен прямо по месту
    lazy var clockUpdate: (Bool) -> Void = {
        // вместо __weak __typeof используем capture list
        // слабая ссылка нужна, т.к. ViewController через свойство clockUpdate
        // ссылается на замыкание, а замыкание ссылается на него
        [weak self] (animated: Bool) in
        
        let units = Set<Calendar.Component>([.hour, .year, .minute, .second])
        let date = NSDate(timeIntervalSinceNow: 1.0)
        let components = NSCalendar.current.dateComponents(units, from: date as Date)
        
        // optional chaining: если self == null, то метод вызыван не будет
        self?.digitViews[0].setDigit(value: uint(components.hour! / 10), animated: animated)
        self?.digitViews[1].setDigit(value: uint(components.hour! % 10), animated: animated)
        self?.digitViews[2].setDigit(value: uint(components.minute! / 10), animated: animated)
        self?.digitViews[3].setDigit(value: uint(components.minute! % 10), animated: animated)
        self?.digitViews[4].setDigit(value: uint(components.second! / 10), animated: animated)
        self?.digitViews[5].setDigit(value: uint(components.second! % 10), animated: animated)

        let time = CFAbsoluteTimeGetCurrent()
        var delta = ceil(time) - time - DigitView.animationDuration
        
        while (delta < 0.0) {
            delta += 1.0
        }
        
        let dtime = DispatchWallTime.now() + DispatchTimeInterval.seconds(Int(delta))
        DispatchQueue.main.asyncAfter(wallDeadline: dtime, qos:.unspecified, flags: [], execute: {
            if let cb = self?.clockUpdate {
                cb(true)
            }
        })

    }
    
    override var prefersStatusBarHidden: Bool {
        return true
    }

}
